052-connector-state.md

```diff
@@ -0,0 +1,77 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Support stopping connectors
+
+This proposal aims at adding support for stopping Kafka Connect connectors.
+
+## Current situation
+
+At the moment Strimzi allows pausing connectors. This is done by a property, `pause`, in the `KafkaConnectorSpec` schema. Whenever it is set to `true`, the connectors and its tasks are paused.
+
+For example:
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  tasksMax: 1
+```
+
+## Motivation
+
+Since Kafka 3.5, Kafka Connect connectors can also be stopped. This is done by using a new REST endpoint `PUT /connectors/{connector}/stop`. Compared to paused where the connector and tasks stay instantiated, when stopped only the configuration of the connector is kept but nothing is actually running.
+
+A connector paused still uses some resources but it's faster to resume. So using pause is great if you want to halt data flowing for a short duration. For longer durations it's interesting to fully stop the connector with the new API. Also the stopped state will be required to use the offset delete/reset endpoints that should come with Kafka 3.6. For these reasons, Strimzi should support both pausing and stopping connectors.
+
+The `PUT /connectors/{connector}/resume` REST endpoint is used to restart both paused and stopped connectors.
+
+## Proposal
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
```

Comment:
Using a field in the spec let me think the different approach we took with the `KafkaRebalance` resource where we are using an annotation for pause, stop, refresh ...
Are we sure we want to go in a different direction here? 
Or we should change the `KafkaRebalance` to have the same approach?
It's more as a general discussion about setting a common pattern for the users from a UX pov about how to interact with custom resources which "run" operations which could be paused, stopped, restarted and so on.

Reply:
I do not think this is the same. AFAIK most of the rebalance annotations are commands. This is a declarative management of the state.

Reply:
I agree with Jakub, the KafkaRebalance annotations are more commands to ask the operator to do something, rather than describing the current expected state of the resource.

In terms of the values, given the new field is called `state`, should the values be `paused`, `stopped`, and `running`

Reply:
Not sure, isn't it a command for the operator to invoke an API to Kafka Connect and ask to run an action (pause, stop and re-start a connector)?

Reply:
No, it describes the desired state. Compare it for example with `strimzi.io/rebalance=refresh` => you are nto asking it to be refreshed every reconciliation (ie. constantly refreshing) but just once.

Reply:
I'm happy to use `paused`, `stopped` and `running` as values. Then it's exactly the same as the `status.connectorStatus.connector.state` field which show the current state (apart as it's in uppercase in this field).

Reply:
@ppatierno @scholzj @tombentley what was your thought on changing the proposed values to `paused`, `stopped` and `running`, rather than `pause`, `stop` and `run`? I think the former makes more sense

Reply:
I'm fine with `paused`, `stopped` and `running`. It should describe the desired state, so it probably fits better.

Reply:
+1 from me as well.

Reply:
While I can't think of anywhere else that I've seen the gerund in resource-based APIs, I don't have a strong opinion. 

TBH the word `state` bothers me more. While it comes from the terminology used by connect itself, it's a bit more overloaded in a kube context, where we sometimes talk about the "desired state" (as a synonym for `spec`), and also use `status` as a way of communicating the connector's "state" more generally. I can't think of anything that is unambiguously better though. 
* Synonyms like `disposition` lose the connection to standard connect terminology
* Something like `connectorState` could be confused for a "KafkaConnector's `status`" . 
* Something like `targetState` might work (it's the term used in the KIP) but doesn't feel entirely natural because the `spec` is always expressing what's desired.

Reply:
I agree that the word `state` isn't ideal, however I do like that it aligns with the Connect output from the REST API and there isn't anything else that comes to mind as being a better fit.